include/uk/plat/paging.h

```diff
@@ -0,0 +1,279 @@
+	__vaddr_t pt_vbase;
+	__paddr_t pt_pbase;
+
+	struct uk_falloc *fa;
+
+	/* Architecture-dependent part */
+	struct ukarch_pagetable arch;
+
+#ifdef CONFIG_PAGING_STATS
+	unsigned long nr_lx_pages[PT_LEVELS];
+	unsigned long nr_lx_splits[PT_LEVELS];
+	unsigned long nr_pt_pages[PT_LEVELS];
+#endif /* CONFIG_PAGING_STATS */
+};
+
+/* Protection flags */
+#define PAGE_PROT_NONE		0x00 /* Page is not accessible */
+#define PAGE_PROT_READ		0x01 /* Page is readable */
+#define PAGE_PROT_WRITE		0x02 /* Page is writeable */
+#define PAGE_PROT_EXEC		0x04 /* Page is executable */
+
+#define PAGE_PROT_RW		(PAGE_PROT_READ | PAGE_PROT_WRITE)
+#define PAGE_PROT_RWX		(PAGE_PROT_RW | PAGE_PROT_EXEC)
+
+/* Page operation flags */
+#define PAGE_FLAG_KEEP_PTES	0x01 /* Preserve PTEs on map/unmap */
+#define PAGE_FLAG_KEEP_FRAMES	0x02 /* Preserve frames on unmap */
+#define PAGE_FLAG_FORCE_SIZE	0x04 /* Force the page size specified with \
+				      * PAGE_FLAG_SIZE() \
+				      */
+
+#define PAGE_FLAG_SIZE_SHIFT	4
+#define PAGE_FLAG_SIZE_MASK	((1UL << PAGE_FLAG_SIZE_SHIFT) - 1)
+
+#define PAGE_FLAG_SIZE(lvl)					\
+	(((lvl) & PAGE_FLAG_SIZE_MASK) << PAGE_FLAG_SIZE_SHIFT)
+
+#define PAGE_FLAG_SIZE_TO_LEVEL(flag)				\
+	(((flag) >> PAGE_FLAG_SIZE_SHIFT) & PAGE_FLAG_SIZE_MASK)
+
+/* Page table clone flags */
+#define PAGE_FLAG_CLONE_NEW	0x01 /* Create an empty page table */
+
+/**
+ * Returns the active page table (the one that defines the virtual address
+ * space at the moment of the execution of this function).
+ */
+struct uk_pagetable *ukplat_pt_get_active(void);
+
+/**
+ * Switches the active page table to the specified one.
+ *
+ * @param pt the page table instance to switch to. The code of the function
+ *   must be mapped into the new address space at the same virtual address.
+ *
+ * @return 0 on success, a non-zero value otherwise
+ */
+int ukplat_pt_set_active(struct uk_pagetable *pt);
+
+/**
+ * Initializes a new page table from the currently configured page table
+ * in hardware and assigns the given physical address range to be available
+ * for allocations and mappings.
+ *
+ * @param pt pointer to an uninitialized page table that will be set to the
+ *   page table hierarchy currently configured in hardware.
+ * @param start start of the physical address range that will be available
+ *   for allocations of physical memory (e.g., mapping with __PADDR_ANY).
+ *   The function may reserve some memory in this area for own purposes. The
+ *   range must not be assigned to other page tables.
+ * @param len the length (in bytes) of the physical address range.
+ *
+ * @return 0 on success, a non-zero value otherwise.
+ */
+int ukplat_pt_init(struct uk_pagetable *pt, __paddr_t start, __sz len);
+
+/**
+ * Adds a physical memory range to the frame allocator of the given page table.
+ * The physical memory is available to all page tables sharing the same frame
+ * allocator.
+ *
+ * @param start start of the physical address range that will be available
```

Comment:
@marcrittinghaus , If I understood correctly, I see some potential problem here. If the function can reserve (use) some of the memory area declared in the arguments to put its metadata, this will forbid us to declare specific chunks for DMA. For instance, in some platforms we might have not only one contiguous physical memory chunk to use, but we might need to declare several chunks because some memory in the middle might be reserved, or for example some specific chunks might be used only for DMA. For instance, let's say we have the 3 memory chunks (AB, BC, CD) illustrated below, and BC is used exclusively for DMA, which means that devices will access this chunk directly (unless we use SMMU for translation). When we declare BC it's suppose that the device should be able to use the whole BC physical range, so we cannot have any metadata inside. 

```
|    -----C1------- | ----C2 DMA area ----- | ----------C3-------------- |
A                      B                                  C                                 D
```

I don't have a better solution to propose, but you could consider the possibility to put metadata into a specific reserved memory inside kernel virtual address space.

Reply:
I see what you mean. Generally, the `addmem` functionality is not supposed to be used to exclude "short" holes. In fact, internally zones are searched linearly in certain circumstances, so there should be as few as possible. The scenario that you described could be solved by defining a zone from A all the way up to D and then immediately do an allocation for C2. This can be done like so:
```C
__paddr_t paddr = START_OF_C2;
pt->fa->falloc(pt->fa, &paddr, LENGTH_OF_C2_IN_FRAMES);
```
Afterwards the memory is never touched again and won't contain metadata. There are no restrictions on the alignment and length of C2.

The allocator itself accepts a pointer to an arbitrary virtual address range to put it's metadata into. It is the x86 helper code that decides to use some of the address range provided to `addmem` to give the allocator for metadata. So you could do it differently if the platform requires it. Only exception is that if the platform provides direct-mapped physical memory (i.e., physical memory is always mapped somewhere in the address space to a fixed offset). Then some of the metadata is placed in the free frames itself to save memory for metadata. For example, to have the range BC in the free list, the allocator might put a list entry at the beginning of BC. Of course, only as long as the memory range is not allocated.

So I think when you allocate the range ahead of time and just "forget" it before initializing the area for DMA it is pretty save to do it this way. It would only be a problem, if there must be a guarantee that there is no memory access between `addmem` and `falloc`. Nevertheless, for this case, I could easily add a config parameter that forbids to put metadata into the frames and always use the user-supplied buffer space, irrespective of the availability of direct-mapped physical memory.